# 🤖 Machine Learning Fundamentals: Regression & Classification

This notebook provides a comprehensive guide to understanding different types of modeling, learning paradigms, and a deep dive into **Linear Regression**.

---

## 📋 1. Types of Modeling

Understanding the nature of your output variable is the first step in choosing a model:

*   **📊 Regression Models**: Used when the target output is a **continuous numerical value** (e.g., predicting house prices, temperature, or stock value).
*   **🏷️ Classification Models**: Used when the output is a **discrete category or label** (e.g., Spam vs. Not Spam, diagnosing a disease).
*   **🔍 Clustering/Recommendation**: Used when there is no predefined label; the model groups data points based on **similarity**.

# Types of Modelling -

- Regression Models -> where the output is continuous ( predicting the price of a product )
- Classification Models -> where the output is a category / label ( discrete output ) - predicting cancer in a patient
- Recommendation/Clustering Models -> where there is no output upfront, based on similarity, closeness to a group ( which has a preference )

In [ ]:
# Predict the revenue of a company ? - Continuous Numerical Space ( 0 - infinity ) anything
# Regression - output is in a Continuous Numerical Space


# Predict cancer in a patient ? - Discrete Categorical Space ( Yes / No )
# classification - output is in a Discrete Categorical Space

# 2 possibility - Binary Classification
# >2 possibility - Multi-Class Classification
# Multi label classification -> In the same prediction - multiple predictions


# Clustering Model -> Output space does not pre-exist, we create a proxy output
# bank customers
# Bank needs to segregate their customer base into groups based on similarity
# so they can sell curated products to the individual groups based on characteristics


# Types of Learning -

- Supervised - Regression, Classification
- Unsupervised - Recommendation Model
- Semisupervised - Some data available from the past - you use it for supervision and then predict for the remaining data for which you don't have the historical data


# Types of Relationships -

- Linear Relationship - straight line fitting - y = mx + c
- Non Linear Relationship - y is not directly related - but has a equation ( non-linear ) between the 2 - y = 2*x^2 + 5 , y = e^x
- Complex relationships - somewhere linear, somewhere non linear, mix of the two in different positions

# Regression Models -

1. Linear Regression Model
2. Multi Linear Regression Model
3. Polynomial Regression Model

# <center>Linear Regression

Linear regression is a supervised learning algorithm that models the relationship between one or more independent variables (predictors) and a dependent variable (outcome). It assumes that this relationship is linear, meaning the change in the dependent variable is proportional to the change in the independent variables.

![](https://www.biomedware.com/files/documentation/spacestat/Statistics/Multivariate_Modeling/Regression/regression_line.png)

### Simple Linear Regression:
Simple linear regression is used when there is one independent variable $x$ and one dependent variable $y$. The goal is to fit a straight line that best represents the relationship between these two variables.

The mathematical form of simple linear regression is:
$$y = \beta_0 + \beta_1 x + \epsilon$$
Where:
- $y$ is the dependent variable (outcome),
- $x$ is the independent variable (predictor),
- $\beta_0$ is the **intercept** (the value of $y$ when $x = 0$),
- $\beta_1$ is the **slope** (how much $y$ changes for a one-unit change in $x$),
- $\epsilon$ is the **error term** (the difference between actual and predicted $y$).



### Example: Predicting Building Height from the Number of Stories

Let's consider the relationship between the number of stories a building has and its height. For this example, we assume that every story adds a constant amount to the building's height. In reality, this could be approximately true if each story is about the same height (say, around 3 meters per story).

#### Step 1: Problem Setup
We want to predict the height of a building based on the number of stories it has. In this case:
- **Independent variable (X)**: The number of stories in the building.
- **Dependent variable (Y)**: The height of the building in meters.

### Step 2: Assumptions
In linear regression, we assume that the relationship between the number of stories and the building's height is linear. This means that for each additional story, the height increases by a constant amount. For example:
- A building with 5 stories might be around 15 meters tall.
- A 10-story building might be around 30 meters tall.
- A 20-story building might be around 60 meters tall.

The constant increase per story represents the slope of the linear relationship. Here, we assume that each story adds approximately 3 meters to the building's total height.

### Step 3: Equation of the Line
In linear regression, we model the relationship between the number of stories and the height using the following equation:
$$Y = \beta_0 + \beta_1 X$$
Where:
- $Y$ is the height of the building (in meters).
- $X$ is the number of stories.
- $\beta_0$ is the intercept (the predicted height of a building when it has 0 stories, which might be 0 or a small value representing base construction height).
- $\beta_1$ is the slope (the increase in height for each additional story).

### Step 4: Fitting the Model
Using historical data of different buildings with known heights and number of stories, we can fit a simple linear regression model. The model will estimate the parameters:
- $\beta_0$, the intercept.
- $\beta_1$, the slope (in this case, the height added per story).

For example, if we observe that a building with 5 stories is 15 meters tall, and a 10-story building is 30 meters tall, the model will find that each additional story adds roughly 3 meters to the height.

### Step 5: Interpretation
Once the model is trained, we can interpret the results:
- **Intercept ($\beta_0$)**: If the intercept is 0, this means that a building with 0 stories has 0 height (which makes sense in this case). If it’s a small positive value, it may represent the base construction height before stories are added.
- **Slope ($\beta_1$)**: The slope would tell us the height added per story. If $\beta_1 = 3$, this means each additional story adds 3 meters to the height.

### Step 6: Using the Model for Prediction
After the model is built, we can use it to predict the height of buildings with any number of stories. For instance:
- A 25-story building would be predicted to have a height of approximately $25 \times 3 = 75$ meters.
- A 40-story building would be predicted to have a height of approximately $40 \times 3 = 120$ meters.

### Multiple Linear Regression:
In multiple linear regression, more than one independent variable is used to predict the dependent variable. For example, predicting sales based on TV, radio, and newspaper advertising expenditures:
$$y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \ldots + \beta_n x_n + \epsilon$$
Where:
- $x_1, x_2, \ldots, x_n$ are the independent variables,
- $\beta_0, \beta_1, \ldots, \beta_n$ are the coefficients.

# How Does Linear Regression Work?

The key idea behind linear regression is to find the line (or hyperplane) that best fits the data points. This is done using the **Ordinary Least Squares (OLS)** method, OLS is a classical method to estimate coefficients in linear regression, but there are other methods like Gradient Descent, MLE, and regularized methods such as Ridge and Lasso.

![](https://www.biomedware.com/files/documentation/spacestat/Statistics/Multivariate_Modeling/Regression/regression_line.png)

## Ordinary Least Squares (OLS) Method:
OLS estimates the coefficients of the regression model ($\beta_0, \beta_1, \ldots$) by minimizing the sum of the squared residuals (errors). A residual is the difference between the actual values $y_i$ and the predicted values $\hat{y_i}$:
$$\text{Residual}_i = y_i - \hat{y}_i$$
The goal of OLS is to minimize the **sum of squared residuals (SSR)**:
$$SSR = \sum_{i=1}^{n} (y_i - \hat{y}_i)^2$$
<img src="https://i.ibb.co/rvGMQqW/0-q-CJ4-Iiy3-Eebo-KF7u.gif" alt="0-q-CJ4-Iiy3-Eebo-KF7u" border="0">

## Gradient Descent

Gradient Descent is an optimization algorithm used for finding the minimum of a function iteratively. In the context of linear regression, Gradient Descent can be used to minimize the cost function (e.g., the sum of squared residuals) without solving the system of equations directly.

Gradient Descent is particularly useful for large datasets or when the normal equations in OLS become computationally expensive to solve due to matrix inversions. It works by updating the coefficients iteratively:
$$\beta_j := \beta_j - \alpha \frac{\partial J(\beta)}{\partial \beta_j}$$
Where:
- $\alpha$ is the learning rate.
- $J(\beta)$ is the cost function (typically the sum of squared errors in linear regression).

**Stochastic Gradient Descent (SGD)** is a variation where the updates are made for individual training examples rather than the whole dataset.

<img src="https://i.ibb.co/SyScPHZ/Gradient-Descent-Animation.gif" alt="Gradient-Descent-Animation" border="0">

## Assumptions of Linear Regression:

Linear regression relies on several assumptions:

### Linearity:
The relationship between the independent and dependent variables should be linear.

### Independence:
Residuals should be independent of each other. The residual for one observation should not depend on the residuals of other observations.

### Homoscedasticity:
The variance of the residuals should be constant across all levels of the independent variables.

### Normality:
The residuals should be normally distributed.

### No Multicollinearity (for Multiple Regression):
In multiple regression, independent variables should not be highly correlated with each other.

## Limitations of Linear Regression:

While linear regression is powerful, it has certain limitations:
- **Linearity Assumption**: If the relationship is not linear, the model won’t capture the true relationship.
- **Outliers**: Linear regression is sensitive to outliers, which can skew the results.
- **Multicollinearity**: In multiple regression, highly correlated independent variables can make the model unstable.
- **Overfitting**: Using too many variables can lead to overfitting, where the model fits the training data too closely but performs poorly on new data.

## Applications of Linear Regression:

Linear regression is widely used in various fields:
- **Economics**: Predicting GDP, inflation, or consumer behavior.
- **Marketing**: Predicting sales based on advertising spending.
- **Healthcare**: Estimating the impact of treatments or drug dosage on patient outcomes.
- **Finance**: Predicting stock prices or credit risk.

Why are we learning linear regression?
- widely used
- runs fast
- easy to use (not a lot of tuning required)
- highly interpretable
- basis for many other methods


# Univariate (Simple) Linear regression
Linear regression with a single variable or feature is called **univariate linear regression**. The output of linear regression is an estimate of the outcome variable (aka target).

In [ ]:
import warnings
warnings.filterwarnings('ignore')#ignore warning printed during the session
import pandas as pd#data handling
import numpy as np#data handling

In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [ ]:
%cd /content/drive/My Drive/UpGrad/Linear Regression/

/content/drive/My Drive/UpGrad/Linear Regression


### Either way you can import the data from your local system

In [ ]:
data = pd.read_csv("Advertising.csv")
data.head()

,TV,Radio,Newspaper,Sales
0,230.1,37.8,69.2,22.1
1,44.5,39.3,45.1,10.4
2,17.2,45.9,69.3,9.3
3,151.5,41.3,58.5,18.5
4,180.8,10.8,58.4,12.9


In [ ]:
#Step 2: select independent(X) and dependent variable (y)
# create a Python list of feature names
input_feature_names = ['TV']

# use the list to select a subset of the original DataFrame
X = data[input_feature_names]

# sales
y = data['Sales']

In [ ]:
# Step 3: Splitting X and y into training and testing sets # train_size = 0.7, test_size =0.3
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    random_state = 42, # makes sure, the splits are not random, repeatable
                                                    train_size=0.80,
                                                    shuffle = True) # Time Series Data

In [ ]:
X_train.head(2)

,TV
79,116.0
197,177.0


In [ ]:
X_train.shape

(160, 1)

In [ ]:
X_test.head(2)

,TV
95,163.3
15,195.4


In [ ]:
X_test.shape

(40, 1)

In [ ]:
from sklearn.linear_model import LinearRegression
# Step 4: Fit linear regression model to trainingset
# Linear Regression Model
linreg = LinearRegression() # inititlaising an object  (linreg) of the Linear Regression Class
# Y = aX + b

# fit the model to the training data (learn the coefficients)
linreg.fit(X_train, y_train)

# y_train = aX_train + b
# solving for a and b - So I get the final eqaution with some values of a and b
# y = 0.0492*X + 6.799

LinearRegression()

In [ ]:
y_pred = linreg.predict(X_test)

In [ ]:
y_pred

array([14.71794394, 16.2115484 , 20.74819743,  7.66403631, 17.37013877,
       10.61402143, 17.2072847 ,  9.44612512, 17.46785121, 15.2669948 ,
        8.58532504,  9.73460946, 18.03086098,  7.37089899, 13.61053628,
       15.03899911,  7.45930549, 16.31391381, 10.62332738, 18.16579721,
       17.79821232, 10.27435438,  8.88776831, 18.79394862, 10.33019006,
        9.60897918, 17.05373658, 13.60123034, 11.17703121,  7.51048819,
       16.41627922, 10.339496  , 16.37440246,  7.90599093, 20.18053468,
       18.203021  ,  9.36702457, 19.85482655, 12.57292322,  8.39920611])

### Mean Absolute Error (MAE):
MAE is the average of the absolute differences between actual and predicted values:

$$
MAE = \frac{1}{n} \sum |y_i - \hat{y}_i|
$$

MAE is less sensitive to outliers than MSE.

In [ ]:
from sklearn import metrics
# Regression Problem - Regression Metrics
metrics.mean_absolute_error(y_test, y_pred)#LOWER THE ERROR, BETTER IS THE MODEL

2.444420003751042

In [ ]:
print("Intercept=",linreg.intercept_)
print("Slope=",linreg.coef_)


Intercept= 7.119638430592953
Slope= [0.04652973]


In [ ]:
# sales = tv_spend*0.04652 + 7.119638
#Weight(Beta):if TV Spend changes by 1$, Sales will changes by 0.05$ keeping all things constant
#Intercept: Expected Sales when TV Spend is 0 is 7.1$

### 🧪 Model Evaluation: Understanding the Metrics

To know if our regression model is performing well, we use several key metrics:

1.  **MAE (Mean Absolute Error)**: The average 'distance' between the actual value and the prediction. Easy to interpret.
2.  **MSE (Mean Squared Error)**: Squares the errors before averaging. This penalizes large errors more heavily.
3.  **RMSE (Root Mean Squared Error)**: The square root of MSE. It brings the error unit back to the original scale (e.g., Dollars instead of Dollars²).
4.  **R² Score (Coefficient of Determination)**: Represents the proportion of variance for the dependent variable that's explained by the independent variables. Range: $(-\infty, 1]$.

In [ ]:
# Evaluate the performance of the model
# Regression Metrics

# Metrics to do this-
# 1. R2 Score -> how much of the change in output, is incorporate by the change in inputs  (-inf to +1)
# 2. MAE  - Mean absolute Error
# 3. MSE - Mean Squared Error - units - squaared, Rs -> Rs^2
# 4. RMSE - Root Mean Squared Error - Original Unit Rs^2 -> Rs
# 5. MAPE - Mean Absolute Percentage Error - Comparison and relatable to all scenarios

In [ ]:
# sklearn metrics -
# R2 score
# MAE
# RMSE

### Mean Squared Error (MSE):
MSE is the average of the squared residuals (errors):

$$
MSE = \frac{1}{n} \sum (y_i - \hat{y}_i)^2
$$

Lower MSE indicates a better fit.

In [ ]:
mse = np.mean((y_test - y_pred) ** 2)
print("Mean Squared Error:", mse)

Mean Squared Error: 10.204654118800956


### Root Mean Squared Error (RMSE):
RMSE is the square root of MSE and provides an error measure in the same units as the dependent variable:

$$RMSE = \sqrt{\frac{1}{n} \sum (y_i - \hat{y}_i)^2}$$

In [ ]:
np.sqrt(metrics.mean_squared_error(y_test, y_pred)) # RMSE

np.float64(3.194472431998898)

After prediction a data scientist should always show how was the performance of the model using different evaluation metrics. One of the most important evaluation metric for linear regression is the coefficient of determination also known as R-squared value ($ R^2 $). The $ R^2 $ is basically a statistical measure of how close the data are to the fitted regression line, by "data" it means our independent variable data, the **engine-size** data. Usually, the larger the $R^2$ value is, the better the model fits the data.

In [ ]:
# R2 - How much of the variation in output is being incorporated by the input
# Is there an input justifying that variation in output
# 2 rows - which are identical
# but the output for the same are different

In [ ]:
# When my input ( TV spends ) is increasing OAPD and the output is also increasing OTPD
# OAD, my input increases, but my Output decreases, keeping everything else constant
# A high (close to 1) R2 score indicates that this isn't happenining

$ R^2 $ is defined as :

\begin{align}
r^2 = \frac{SS_{regression}}{SS_{total}}
\end{align}

\begin{align}
r^2 = 1 - \frac{SS_{error}}{SS_{total}}
\end{align}

Where:
 - $ SS_{regression} $ - "regression sum of squares"; quantifies how far the predicted value is from the mean
 - $ SS_{error} $ - "error sum of squares"; quantifies the variance of the target data points around the regression line
 - $ SS_{total} $ - "total sum of squares"; quantifies the variance of the target data points around the mean

\begin{align}
SS_{regression} = \sum_{i=1}^m(\hat y_i - \bar y)^2
\end{align}

\begin{align}
SS_{error} = \sum_{i=1}^m(y_i - \hat y)^2
\end{align}

\begin{align}
SS_{total} = \sum_{i=1}^m(y_i - \bar y)^2
\end{align}

In [ ]:
ss_total = np.sum((y_test - np.mean(y_test)) ** 2)  # Total sum of squares
ss_residual = np.sum((y_test - y_pred) ** 2)  # Residual sum of squares
r2 = 1 - (ss_residual / ss_total)
print("R-squared:", r2)

R-squared: 0.6766954295627076


In [ ]:
# r2_score -> -infinity to +1
# R2 can't be negative -> How much of variation in the ouput is explained by the input ?
# If there are 5 factors affecting sales - and you take all 5 as input to the model - there is a high probability that you
# will get good r2 - but if you take 3 inputs where 5 are needed - obviously the inputs are not sufficient to explain the output

# Your model is predicting reverse of what the actual trend is - r2 is negative - reverse explnation of output based on input

### Interpreting Model Coefficients

How do we interpret the TV coefficient ($\beta_1$)?
- A "unit" increase in TV ad spending is **associated with** a 0.0492751 "unit" increase in Sales.
- Or more clearly: An additional $1,000 spent on TV ads is **associated with** an increase in sales of 49.2751 widgets.

Note that if an increase in TV ad spending was associated with a **decrease** in sales, $\beta_1$ would be **negative**.

In [ ]:
# Final Equation :  y = 0.049*(X+1) + 6.7999 = 0.049X + 0.049 + 6.7999
# y = 0.049*X + 6.7999
# Baseline ->
# Dominos stops advertising today - Even then - they'll make some sales right - brand value - baseline sales
# No, it will keep on decreasing  - baseline also has a loss function associated
# Short Term and Long Term ROI - You have 100 rs today , I have 50 Rs Today

# we both saw an advertisement for a product worth 80 Rs - you bought it immediately - Short Term ROI
# I saved for another week - I bought it then(after a week) - Long Term ROI

In [ ]:
# y = 0.04893566*TV_spends + 7.0687776

# TV_spends=0 ( No marketing in TV ) -> 7.06877 Baseline sales
# For every unit increase in the input variable TV - my output variable increase by - 0.048

# y = 1.5x + 5
# y` = 1.5(x+1) + 5
# y` - y = 1.5(x+1) + 5 - (1.5x + 5)
# incremental value =  1.5

# Multiple Linear Regression:
In multiple linear regression, more than one independent variable is used to predict the dependent variable.

In [ ]:
# create a Python list of feature names
input_feature_names = ['TV', 'Radio', 'Newspaper']

# use the list to select a subset of the original DataFrame
X = data[input_feature_names]

# sales
y = data['Sales']

In [ ]:
# Splitting X and y into training and testing sets # train_size = 0.7, test_size =0.3
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    random_state = 42, # makes sure, the splits are not random, repeatable
                                                    train_size=0.80,
                                                    shuffle = True) # Time Series Data

In [ ]:
# Fit linear regression model to trainingset
# Linear Regression Model
linreg = LinearRegression() # inititlaising an object  (linreg) of the Linear Regression Class

# fit the model to the training data (learn the coefficients)
linreg.fit(X_train, y_train)

LinearRegression()

In [ ]:
y_pred = linreg.predict(X_test)

In [ ]:
# Calculate the R² score
r2 = metrics.r2_score(y_test, y_pred)

print("R-squared:", r2)

R-squared: 0.899438024100912


In [ ]:
print("Intercept=",linreg.intercept_)
print("Slope=",linreg.coef_)


Intercept= 2.979067338122629
Slope= [0.04472952 0.18919505 0.00276111]


## Feature Selection

In [ ]:
# There can be lots of variable that impact my output, macroeconomic, microeconomic, geographical, political, etc....
# should I be considering all of them ?
# Step 1 : Things that are in our control
# Step 2 : I cannot change it very frequently, plus it might not have a very significant impact compared to the cost of controlling it
# Step 3 : How much impact / significance does this variable have ?  - assume

# We must find this out mathematically
# Variability of the output to the variability of the input

## Hypothesis Testing and p-values

Generally speaking, you start with a **null hypothesis** and an **alternative hypothesis** (that is opposite the null). Then, you check whether the data supports **rejecting the null hypothesis** or **failing to reject the null hypothesis**.

(Note that "failing to reject" the null is not the same as "accepting" the null hypothesis. The alternative hypothesis may indeed be true, except that you just don't have enough data to show that.)

As it relates to model coefficients, here is the conventional hypothesis test:
- **null hypothesis:** There is no relationship between TV ads and Sales (and thus $\beta_1$ equals zero)
- **alternative hypothesis:** There is a relationship between TV ads and Sales (and thus $\beta_1$ is not equal to zero)

How do we test this hypothesis? Intuitively, we reject the null (and thus believe the alternative) if the 95% confidence interval **does not include zero**. Conversely, the **p-value** represents the probability that the coefficient is actually zero:

In [ ]:
# Significant relationship between input and output or not
# p-value of an input feature <0.05 -> reject the null hypothesis, there is a significant relationship between input and output
# p-value -> >0.05 - there is no significant relationship

# P low ( <0.05 ), null go or P 1 null stay

In [ ]:
from sklearn.feature_selection import f_regression
fregression = f_regression(X_train, y_train) #returns Fvalues of features; p values of features
fregression

(array([228.46566066,  85.41607389,   9.47648406]),
 array([1.68406821e-32, 1.57312319e-16, 2.45446054e-03]))

a p-value less than 0.05 is one way to decide whether there is likely a relationship between the feature and the response. (Again, using 0.05 as the cutoff is just a convention.)

In this case, the p-value for TV is far less than 0.05, and so we **believe** that there is a relationship between TV ads and Sales.

Note that we generally ignore the p-value for the intercept.

In [ ]:
print("Intercept=",linreg.intercept_)
print("Slope=",linreg.coef_)

In [ ]:
print("RMSE=", np.sqrt(metrics.mean_squared_error(y_test, y_pred)))
print("Rsquare=",metrics.r2_score(y_test,y_pred))

### Adjusted R-squared statistic

The Adjusted R-squared takes into account the number of independent variables used for predicting the target variable. In doing so, we can determine whether adding new variables to the model actually increases the model fit.

Let’s have a look at the formula for adjusted R-squared to better understand its working.

\begin{align}
AdjustedR^2 = {1-\frac{(1-r^2)(n-1)}{(n-k-1)}}
\end{align}

Where,

    n represents the number of data points in the dataset
    k represents the number of independent variables, and
    r represents the R-squared values determined by the model.


So, if R-squared does not increase significantly on the addition of a new independent variable, then the value of Adjusted R-squared will actually decrease.

On the other hand, if on adding the new independent variable we see a significant increase in R-squared value, then the Adjusted R-squared value will also increase.

In [ ]:
# Number of observations (n) and number of predictors (p)
n = len(y_test)  # Number of samples
p = X.shape[1]  # Number of independent variables (predictors)

# Calculate adjusted R-squared
adj_r2 = 1 - (1 - r2) * (n - 1) / (n - p - 1)
print("R-squared:", r2)
print("Adjusted R-squared:", adj_r2)

# Checking the The assumptions of Linear Regression
Those are crucial for ensuring that the model is valid and the results are reliable. You can check these assumptions using various diagnostic techniques.

## Linearity:<br>
The relationship between the independent and dependent variables should be linear.

**How to Check:**<br>
Plot the residuals (errors) against the predicted values. If the relationship is linear, the residuals should be randomly scattered around zero.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Set modern theme
sns.set_theme(style="whitegrid", palette="muted")

# Residuals Calculation
residuals = y_test - y_pred

# Modern Residual Plot
plt.figure(figsize=(10, 6))
sns.residplot(x=y_pred, y=residuals, lowess=True,
             line_kws={'color': 'red', 'lw': 2, 'alpha': 0.8})

plt.xlabel('Predicted Sales', fontsize=12)
plt.ylabel('Residuals (Actual - Predicted)', fontsize=12)
plt.title('Residuals vs. Predicted Values (Linearity & Bias Check)', fontsize=14, fontweight='bold')
plt.tight_layout()
plt.show()

- If the scatter plot shows no clear pattern and the points are randomly distributed, the linearity assumption holds.

## Independence of Errors:
The residuals should be independent of each other. This is particularly important for time-series data.

**How to Check:**<br>
Plot residuals against time or use statistical tests like the Durbin-Watson test.

In [ ]:
from statsmodels.stats.stattools import durbin_watson

# Durbin-Watson Test
dw = durbin_watson(residuals)
print(f"Durbin-Watson Statistic: {dw}")

- A value close to 2.0 indicates that there is no autocorrelation, while values less than 1.5 or greater than 2.5 suggest correlation.

## Homoscedasticity:
The variance of the residuals should be constant across all levels of the independent variables. In other words, the residuals should have the same spread for all predicted values.

**How to Check:**<br>
Plot the residuals against the predicted values. If the variance is constant, the residuals should be evenly spread around zero without a funnel shape.

In [ ]:
plt.figure(figsize=(10, 6))
sns.scatterplot(x=y_pred, y=residuals, color='#55a868', s=80, alpha=0.7)
plt.axhline(y=0, color='#c44e52', linestyle='--', linewidth=2)

plt.xlabel('Predicted Values', fontsize=12)
plt.ylabel('Residuals', fontsize=12)
plt.title('Homoscedasticity Check: Variance of Errors', fontsize=14, fontweight='bold')
plt.grid(True, linestyle=':', alpha=0.7)
plt.tight_layout()
plt.show()

- If you see a funnel shape (i.e., increasing or decreasing spread of residuals as predicted values change), then there may be a problem with homoscedasticity.

## Normality of Residuals:
The residuals should be normally distributed. This can be checked by plotting a histogram or using a Q-Q plot (quantile-quantile plot).

**How to Check:**<br>
Use a histogram or Q-Q plot of the residuals.

In [ ]:
import seaborn as sns
import scipy.stats as stats
import matplotlib.pyplot as plt

fig, axes = plt.subplots(1, 2, figsize=(15, 6))

# 1. Distribution of Residuals
sns.histplot(residuals, kde=True, color="#4C72B0", ax=axes[0])
axes[0].set_title('Distribution of Error Terms', fontsize=14, fontweight='bold')
axes[0].set_xlabel('Residual Value')

# 2. Probability Plot (Q-Q Plot)
stats.probplot(residuals, dist="norm", plot=axes[1])
axes[1].get_lines()[0].set_markerfacecolor('#C44E52')
axes[1].get_lines()[0].set_markersize(8)
axes[1].set_title('Normal Q-Q Plot', fontsize=14, fontweight='bold')

plt.tight_layout()
plt.show()

- If the residuals form a straight line on the Q-Q plot and the histogram is roughly bell-shaped, the normality assumption holds.

## Multicollinearity (For Multiple Linear Regression):
In multiple linear regression, the independent variables should not be highly correlated with each other.

**How to Check:**<br>
Use the Variance Inflation Factor (VIF) to check for multicollinearity. VIF values greater than 5 or 10 indicate potential multicollinearity.

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.tools.tools import add_constant

# Add constant column for VIF calculation not necessary
X_multi_with_const = add_constant(data[input_feature_names])

# Calculate VIF for each feature
vif_data = pd.DataFrame()
vif_data['feature'] = X_multi_with_const.columns
vif_data['VIF'] = [variance_inflation_factor(X_multi_with_const.values, i) for i in range(X_multi_with_const.shape[1])]

print(vif_data)

# Finally checking the predictions

In [ ]:
df_predicted=pd.DataFrame()
df_predicted['Actual']=y_test
df_predicted['Predicted']=y_pred
df_predicted.head()

In [ ]:
y_error = y_pred - y_test
x_range = np.arange(1, len(y_pred) + 1)

plt.figure(figsize=(12, 5))
plt.fill_between(x_range, y_error, color="skyblue", alpha=0.4)
plt.plot(x_range, y_error, color="Slateblue", alpha=0.8, linewidth=2, label='Prediction Error')
plt.axhline(y=0, color='black', linestyle='-', linewidth=0.8)

plt.title("Sequential Error Analysis", fontsize=14, fontweight='bold')
plt.xlabel("Observation Index", fontsize=12)
plt.ylabel("Error Magnitude", fontsize=12)
plt.legend()
plt.grid(axis='y', linestyle='--', alpha=0.6)
plt.show()

In [ ]:
plt.figure(figsize=(12, 6))

x_axis = np.arange(1, len(y_test) + 1)

# Plotting with distinct markers and styles
plt.plot(x_axis, y_test, marker='o', linestyle='-', label='Actual Values', color='#1f77b4', alpha=0.8)
plt.plot(x_axis, y_pred, marker='X', linestyle='--', label='Model Predictions', color='#ff7f0e', alpha=0.9)

plt.title("Model Performance: Actual vs Predicted Sales", fontsize=15, pad=20, fontweight='bold')
plt.xlabel("Observation Index", fontsize=12)
plt.ylabel("Sales Units", fontsize=12)
plt.legend(frameon=True, shadow=True, borderpad=1)
plt.grid(True, linestyle='--', alpha=0.6)
plt.tight_layout()
plt.show()

# Polynomial Regression
Many relationships of interest, however, are not strictly linear. Often we want to model a non-linear relationship—for example, the relationship between the number of hours a student studies and the score she gets on the test. Intuitively, we can imagine there is a big difference in test scores between students who study for one hour compared to students who did not study at all. However, there is a much smaller difference in test scores between a student who studied for 99
hours and a student who studied for 100 hours. The effect one hour of studying has on a student’s test score decreases as the number of hours increases.

What if your data is actually more complex than a simple straight line? Surprisingly, you can actually use a linear model to fit nonlinear data. A simple way to do this is to add powers of each feature as new features, then train a linear model on this extended set of features. This technique is called Polynomial Regression.

In [ ]:
# X - X` ( original independent features - polynomial features )
# y -> y

In [ ]:
# y = ax1 + bx2**2 + cx1x3 + d
# x1, x2, x3  -> Original Form
# y = 2.3x1+ 0.00005*x2, 0.00002*x3, 0.00005*x1^2, 3.4*x2^2, x3^2 , x1x2, x2x3, x3x1 -> Polynomial form ( Max Power = 2)

In [ ]:
#read csv file
data.head()

In [ ]:
# create a Python list of feature names
feature_names=['TV','Newspaper']

# use the list to select a subset of the original DataFrame
X = data[feature_names]

# sales
y = data['Sales']

In [ ]:
X.head(2)

In [ ]:
# x1, x2 -> polynomial features ( of a certain degree : 3 ) -> x1^3, x2^3, x1^2, x2^2, x1, x2, x1*x2 , 1

In [ ]:
from sklearn.preprocessing import PolynomialFeatures
# PolynomialFeatures (prepreprocessing)
poly = PolynomialFeatures(degree = 2) # MUST SPECIFY THE MAXIMUM DEGREE of the Polynomial

# X_poly = poly.fit_transform(X)
# train_test_split(X_poly , y)

X_new = poly.fit_transform(X) # poly.fit(X) , poly.transform(X) => poly.fit_transform(X)

In [ ]:
X_new[0]

In [ ]:
X.loc[0]
# x1, x2 -> 1, x1, x2, x1**2 , x1*x2 , x2**2

In [ ]:
# Splitting X and y into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1,test_size=0.20)

In [ ]:
# target data - what we are predicting - no change / transformation
# X values - X_train and X_test

In [ ]:
# Linear Regression Model
linreg = LinearRegression()

# fit the model to the training data (learn the coefficients)
linreg.fit(X_train, y_train)

In [ ]:
# make predictions on the testing set
y_pred = linreg.predict(X_test)

In [ ]:
# compute the RMSE of our predictions
print("RMSE=",np.sqrt(metrics.mean_squared_error(y_test, y_pred)))
print("Rsquare=",metrics.r2_score(y_test,y_pred))

In [ ]:
df_predicted=pd.DataFrame()
df_predicted['Actual']=y_test
df_predicted['Predicted']=y_pred
df_predicted

In [ ]:
plt.figure(figsize=(12, 6))
indices = np.arange(1, len(y_test) + 1)

plt.plot(indices, y_test, 'o-', label='Actual Sales', markersize=6, color='#2c3e50', alpha=0.7)
plt.plot(indices, y_pred, 's--', label='Polynomial Predicted', markersize=6, color='#e67e22', alpha=0.9)

plt.title("Polynomial Model: Actual vs Predicted (Test Set)", fontsize=15, fontweight='bold', pad=15)
plt.xlabel("Data Point Index", fontsize=12)
plt.ylabel("Sales Units", fontsize=12)
plt.legend(frameon=True, facecolor='white', shadow=True)
plt.grid(True, which='both', linestyle='--', alpha=0.5)
plt.tight_layout()
plt.show()

## 🚀 Summary of Findings

- **Simple Linear Regression**: Used only `TV` as a predictor. It captured the general trend but had higher error.
- **Multiple Linear Regression**: Including `Radio` and `Newspaper` significantly improved the **R² score** (from ~0.67 to ~0.90).
- **Feature Significance**: Using p-values, we observed that `TV` and `Radio` are highly significant, while `Newspaper` often has a p-value > 0.05, suggesting it might not be a strong predictor in this specific dataset.
- **Polynomial Regression**: Allows us to capture non-linear patterns by increasing the degree of our features.

## 💼 Business Insights & Strategic Recommendations

Based on the analysis of the advertising spend and sales data, here are the key takeaways for the marketing department:

### 1. 📈 Channel Efficiency (ROI Analysis)
*   **TV is King**: Every $1,000 increase in TV spending results in approximately **45-49 additional units** of sales. This is your most reliable driver for scale.
*   **Radio's High Impact**: Radio shows a very high coefficient (~0.18). While it often has a smaller total budget than TV, it is highly efficient at driving incremental sales for every dollar spent.
*   **Newspaper Caution**: The p-value for Newspaper was consistently above 0.05. This suggests that **Newspaper advertising does not have a statistically significant impact** on sales in this mix. Consider reallocating the Newspaper budget to TV or Radio to maximize ROI.

### 2. 🎯 Budget Optimization
*   **The Baseline**: The intercept (~2.9 to 7.1) suggests that even with zero advertising, there is a base level of sales (brand equity). Marketing efforts should focus on exceeding this baseline.
*   **Synergy Effects**: The Multiple Linear Regression model's high $R^2$ (0.90) indicates that TV and Radio together explain 90% of the sales variance. This suggests a highly predictable marketing funnel.

### 3. ⚠️ Risk Management
*   **Non-Linearity**: As seen in the Polynomial Regression, there may be 'saturation points' where spending more doesn't lead to the same linear growth. We should monitor for diminishing returns in high-spend channels.
*   **Model Reliability**: With an RMSE of ~1.78 in the multiple model, management can expect sales predictions to be accurate within roughly $\pm$ 1,780 units, providing a solid foundation for supply chain and inventory planning.